OpenRA.Mods.Common/Traits/World/PathFinder.cs

```diff
@@ -82,7 +82,8 @@ public List<CPos> FindUnitPath(CPos source, CPos target, Actor self, Actor ignor
 			if (distance.LengthSquared < 3 && !canMoveFreely)
 				return new List<CPos> { };
 
-			if (source.Layer == target.Layer && distance.LengthSquared < 3 && canMoveFreely)
+			var movementCost = locomotor.MovementCostToEnterCell(self, target, target - source, check, null);
```

Comment:
It looks like you could replace `var distance` above with `var delta = target - source` and then use `delta.LengthSquared` and `delta` directly here.

Reply:
On second look, you really should be overriding `CanMoveFreelyInto` instead of adding an additional check here?

Reply:
See also #17299 for reference. You can do this without touching the pathfinder itself.

Reply:
CanMoveFreelyInto explicitely says its purpose is to check for actors, not for terrain. Sure, using that method would work too, but the goal is terrain / map based walls, not actors blocking. Having actors blocking eachother while going diagonal is out of scope of this pr, as its also a actor based feature (above CanMoveFreelyInto). Sure its an important thing, but this pr is about a posibility to allow custom traits to implement a way to block movement between two directly adjacent cells on a terrain basis, or did i misunderstand #17299 ?

Reply:
For the diagonal  checks diagonal movement the logic should something like "first check X" if that works, then check "Y". if that works too, allow diagonal to XY. Otherwise forbid diagonal. That should solve most of these problems. I might want to do that in another pr, as this logic applies to tiles and actors.

Reply:
Fwiw, I wasn't suggesting you scope creep diagonal blocking in. I was just offering an example of a PR that tried to do something similar without touching the pathfinder.

I also see now that you explicitly want your walls to be part of the terrain and not as wall actors, so I guess that changes things a bit.

Reply:
Thanks for hinting the pr, it hinted me to a combination of both approaches implementic the logic twice: for actors, and for terrain. While my approach did work on the terrain side, i adopted your changes for actor checkings.